src/main/scala/org/apache/spark/sql/execution/datasources/greenplum/GreenplumUtils.scala

```diff
@@ -60,71 +61,169 @@ object GreenplumUtils extends Logging {
+   * @param options Options for the Greenplum data source
+   */
+  def nonTransactionalCopy(
+      df: DataFrame,
+      schema: StructType,
+      options: GreenplumOptions): Unit = {
     df.foreachPartition { rows =>
-      val conn = JdbcUtils.createConnectionFactory(options)()
-      val copyManager = new CopyManager(conn.asInstanceOf[BaseConnection])
+      copyPartition(rows, options, schema, options.table)
+    }
+  }
+
+  /**
+   * Copy a partition's data to a gptable.
+   *
+   * @param rows rows of a partition will be copy to the Greenplum
+   * @param options Options for the Greenplum data source
+   * @param schema the table schema in Greemnplum
+   * @param tableName the tableName, to which the data will be copy
+   * @param accumulator account for recording the successful partition num
+   */
+  def copyPartition(
+      rows: Iterator[Row],
+      options: GreenplumOptions,
+      schema: StructType,
+      tableName: String,
+      accumulator: LongAccumulator = null): Unit = {
+    val valueConverters: Array[(Row, Int) => String] =
+      schema.map(s => makeConverter(s.dataType, options)).toArray
+    val conn = JdbcUtils.createConnectionFactory(options)()
+    val copyManager = new CopyManager(conn.asInstanceOf[BaseConnection])
+
+    try {
+      val tmpDir = Utils.createTempDir(Utils.getLocalDir(SparkEnv.get.conf), "greenplum")
+      val dataFile = new File(tmpDir, UUID.randomUUID().toString)
+      val out = new BufferedOutputStream(new FileOutputStream(dataFile))
       try {
-        val tmpDir = Utils.createTempDir(Utils.getLocalDir(SparkEnv.get.conf), "greenplum")
-        val dataFile = new File(tmpDir, UUID.randomUUID().toString)
-        val out = new BufferedOutputStream(new FileOutputStream(dataFile))
-        try {
-          rows.foreach(r => out.write(convertRow(r)))
-        } finally {
-          out.close()
-        }
-        val in = new BufferedInputStream(new FileInputStream(dataFile))
-        val sql = s"COPY ${options.table}" +
-          s" FROM STDIN WITH NULL AS 'NULL' DELIMITER AS E'${options.delimiter}'"
-        try {
-          logInfo("Start copy steam to Greenplum")
-          val start = System.nanoTime()
-          val nums = copyManager.copyIn(sql, in)
-          val end = System.nanoTime()
-          logInfo(s"Copied $nums row(s) to Greenplum," +
-            s" time taken: ${(end - start) / math.pow(10, 9)}s")
-        } finally {
-          in.close()
+        rows.foreach(r => out.write(
+          convertRow(r, schema.length, options.delimiter, valueConverters)))
+      } finally {
+        out.close()
+      }
+      val in = new BufferedInputStream(new FileInputStream(dataFile))
+      val sql = s"COPY $tableName" +
+        s" FROM STDIN WITH NULL AS 'NULL' DELIMITER AS E'${options.delimiter}'"
+      try {
+        logInfo("Start copy steam to Greenplum")
+        val start = System.nanoTime()
+        val nums = copyManager.copyIn(sql, in)
```

Comment:
ever considering add a timeout for copy?

Reply:
I will do it.

Reply:
maybe another pull request is fine